feat: Automatically retry failed webhook deliveries

jackie-linz opened the pull request:
GitHub webhook deliveries can fail for many reasons. This code goes over all failed deliveries of the past one hour and retries them. This should help avoid cases of jobs pending forever because a runner wasn't provisioned for them. It uses GitHub API to iterate over all deliveries, find the ones that failed, confirm there is no newer delivery that worked, and then finally use GitHub API to redeliver the webhook call.

This only works for app authentication. Personal authentication token has no webhook attached, so it's impossible for us to know which webhook deliveries to monitor. Theoretically, if we ask for enough permission, we could go over all of the user's webhooks (from which repo/org?) and find the matching one based on our webhook URL.

This change adds a Lambda that runs every five minutes. It's an added cost, but should still be basically free. Around 10,000GB/s a month with 14,600 requests. 

Resolves #738

kichik commented:
Might be wrong dependencies. Try running `yarn` again.

jackie-linz commented:
hmm... no luck. I've tried running

```sh
 rm -rf node_modules/ assets/ dist/ lib/
yarn
npm run integ:default:snapshot
```

and there's no change to commit

kichik commented:
I pushed a merge from `main` + redid the snapshot. Hopefully this one works. There are some corner cases with projen's snapshot. It can be a bit brittle at times. I think the last one I had had to do with line breaks? I mostly try to fix them when I find them, so honestly not sure. If you want to dig into it, you can use the jsii docker image and build it there. Then if you finally get the right hash as here, you can diff the template it generated to see what happened.

jackie-linz commented:
I think I found it - the linter is changing the code, which resulted in a diff.

jackie-linz commented:
I did not run `yarn build` before, as jest keeps failing - looks like my machine has too many CPU cores (so jest spawned 6 workers) and not enough memory to match so test keeps timing out and getting killed.

had to add the following to projenrc temporarily to get build running 😆 
```javascript
  jestOptions: {
    jestConfig: {
      maxWorkers: 3,
    },
  },
```

kichik commented:
I went into the code to fix up the logging a bit, ended up having to fix app authentication mode, and then somehow more and more piled on.

The commit message is hopefully descriptive enough. I ran out of time so the unit test is AI slop that does pass but is of questionable value.

Let me know what you think.

jackie-linz commented on src/webhook-redelivery.lambda.ts:
```diff
@@ -0,0 +1,89 @@
+      throw new Error('Failed to fetch webhook deliveries');
+    }
+
+    for (const delivery of response.data) {
```
do you not want to exclude `event`/`action` that we don't care about?

when installed at the org level, apart from `workflow_job` events, this app would also get sent `installation_repositories` with `action` of `added`/`removed`.

the current webhook would return 400 error with `Expecting workflow_job` message for these, so they will get retries for 12 times as it would keep failing

kichik replied:
I don't want to duplicate the logic of which action we care about. If the webhook changes in the future, we will almost definitely forget to update the logic here too. 

But we should definitely fix the success logic to prevent endless redeliveries.

kichik replied:
Are `installation_repositories` events currently marked as failed regardless of how we check `status`/`status_code`? If so, I'm leaning towards returning 200 with `"ignored"` as the message.

jackie-linz reviewed (commented):
I like the new implementation that removed the need to fetch delivery detail.

A few minor comments for your consideration

kichik commented on src/webhook-redelivery.lambda.ts:
```diff
@@ -0,0 +1,89 @@
+
+    for (const delivery of response.data) {
+      const deliveredAt = new Date(delivery.delivered_at);
+      const success = delivery.status_code >= 200 && delivery.status_code < 300;
```
Weird. I thought that field was the text the webhook returned. I didn't want to rely on that as it can technically change. We will have to check what happens when we return something else.

kichik commented:
Thought about it some more and I'm confused by my auth fix. How did it work for you? Were you actually used PAT? Because it seems to be that there will be no webhook access for PAT. It's not attached to an app. The API we use is for apps only. 

This also means redelivery will not work for PAT. I wonder if we can make that one work too somehow. We have no info about the webhook the user creates manually.

jackie-linz commented:
> Thought about it some more and I'm confused by my auth fix. How did it work for you? Were you actually using PAT? Because it seems to be that there will be no webhook access for PAT. It's not attached to an app. The API we use is for apps only.
> 
> This also means redelivery will not work for PAT. I wonder if we can make that one work too somehow. We have no info about the webhook the user creates manually.

I'm not that familiar with different auth for Github apps, but this is my local version of the lambda that worked reusing the existing github secret and github private key secret:

```javascript
import { createAppAuth } from '@octokit/auth-app';
import { Octokit } from 'octokit';

export interface GithubSetting {
  domain: string;
  appId: number;
}

export async function getGithubSetting(): Promise<GithubSetting> {
  if (!process.env.GITHUB_SECRET_ARN) {
    throw new Error('Missing GITHUB_SECRET_ARN environment variable');
  }
  return JSON.parse(await getSecretValue(process.env.GITHUB_SECRET_ARN)) as GithubSetting;
}

export async function getGithubPrivateKey(): Promise<string> {
  if (!process.env.GITHUB_PRIVATE_KEY_SECRET_ARN) {
    throw new Error('Missing GITHUB_PRIVATE_KEY_SECRET_ARN environment variable');
  }
  return await getSecretValue(process.env.GITHUB_PRIVATE_KEY_SECRET_ARN);
}

export const handler = async (): Promise<void> => {
  const setting = await getGithubSetting();
  const privateKey = await getGithubPrivateKey();

  const octokit = new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: setting.appId,
      privateKey,
    },
  });
  ...
```

jackie-linz commented:
deployed and ran successfully twice
![image](https://github.com/user-attachments/assets/1baecdf9-37b6-44b3-b7c7-42a56784fdee)

The log content looks fine too, will keep an eye to compare the behaviour between this and my own one (currently both are actively running)
![image](https://github.com/user-attachments/assets/d450ee88-b3f9-49d9-93aa-b703d9d3d828)

It'd be nice to provide Logs Insight queries like you already do for the others

jackie-linz commented:
Performance comparison

Mine one
![image](https://github.com/user-attachments/assets/8577b21b-7090-493e-9611-1c22bbdd1ca8)

New one
![image](https://github.com/user-attachments/assets/098a30fc-b0f6-4e26-91d5-9129f9bfb7bc)

Looks like the refactored implementation is quite a lot slower, likely due to it needing to fetch 1 hours worth of deliveries instead of 5 mins 😞

kichik commented:
Thanks!

> It'd be nice to provide Logs Insight queries like you already do for the others

Done.

> Looks like the refactored implementation is quite a lot slower, likely due to it needing to fetch 1 hours worth of deliveries instead of 5 mins 😞

My hope was that not calling GitHub API to get delivery information for each delivery would outweigh that. Seems like you have too many deliveries to make that happen 😅

Is the price still basically zero or do we need to bring back the SSM or something similar?

jackie-linz commented:
> My hope was that not calling GitHub API to get delivery information for each delivery would outweigh that. Seems like you have too many deliveries to make that happen 😅

It's more about the extreme low probability of delivery failures than the volume of delivery.

In our setup there's only `10` failures out of `221573` webhook deliveries in the past week, which is less than 0.005%
 
> Is the price still basically zero or do we need to bring back the SSM or something similar?

Price is not that big a concern - average of `24s` per run means `<$0.5` per month

On the other hand, the extended execution time could potentially be a concern, if it exceeds the limit of 4.5 minutes. With our setup it peaked around 54 seconds today, so it is possible for larger orgs to run into problem with the time limit.

![image](https://github.com/user-attachments/assets/ad00fbde-85b3-44c4-af89-1f054d58ef7f)

kichik commented:
I didn't get to test much and unit tests are still AI slop. If you can help me confirm latency is down + redelivery still works as expected, that'd be really appreciated.

jackie-linz commented on src/webhook-redelivery.lambda.ts:
```diff
@@ -23,47 +18,37 @@ async function summarizeDeliveries(octokit: Octokit, timeLimitMs: number) {
+      }
+
+      if (deliveredAt < new Date(Date.now() - 1000 * 60 * 60)) {
+        // stop processing if the delivery is older than 30 minutes (for first iteration and performance)
```
30 or 60 minutes? may want to extract to a global constant and reused at line 66 below?

jackie-linz commented on src/webhook-redelivery.lambda.ts:
```diff
@@ -74,23 +59,42 @@ export async function handler() {
         });
       }
     }
+    lastDeliveryIdProcessed = Math.max(lastDeliveryIdProcessed, details.id);
```
this would not work as expected, as `deliveries` now only include failures, so most of the time it would be empty and `lastDeliveryIdProcessed` would not get updated at all

jackie-linz commented on src/webhook-redelivery.lambda.ts:
```diff
@@ -0,0 +1,128 @@
+        return deliveries;
+      }
+
+      console.log({
```
this is going to generate one log per delivery and would be quite busy. consider change to `console.debug`?

kichik commented:
Thanks. Addressed feedback.

But I also realized `failures` is unbounded. We would need to return successful redeliveries from `newDeliveryFailures` too. Or maybe just TTL items there.

This also makes me think more of a case with too many failed deliveries. Like maybe a misconfigured webhook for a few hours. The Lambda might fail in a loop and reset its cache between each run. It would be nice if we could use the same SQS trick as idle reaper. It returns the message to the queue on failure and uses SQS delivery delay and visibility timeout to ensure the message is retested every 10 minutes. But there is no constant message we can throw back on the queue here and we can't even check if the redelivery was successful without iterating the whole list. There is no delivery GUID -> delivery id / success API.

jackie-linz commented:
> Thanks. Addressed feedback.

👍 

> But I also realized `failures` is unbounded. We would need to return successful redeliveries from `newDeliveryFailures` too. Or maybe just TTL items there.

Lambda typically gets a cold start every few hours, so I don't expect `failures` to grows too much.

If you are worried, we could expire failures with `firstDeliveredAt` that's too old at the end of each run.

> This also makes me think more of a case with too many failed deliveries. Like maybe a misconfigured webhook for a few hours. The Lambda might fail in a loop and reset its cache between each run. It would be nice if we could use the same SQS trick as idle reaper. It returns the message to the queue on failure and uses SQS delivery delay and visibility timeout to ensure the message is retested every 10 minutes. But there is no constant message we can throw back on the queue here and we can't even check if the redelivery was successful without iterating the whole list. There is no delivery GUID -> delivery id / success API.

You mean too many `redeliver` calls can cause the lambda to exceed the 4.5 minute timeout? How about doing `redeliver` in parallel with `Promise.all`?

kichik commented:
> If you are worried, we could expire failures with `firstDeliveredAt` that's too old at the end of each run.

Probably what we will end up doing.

> You mean too many `redeliver` calls can cause the lambda to exceed the 4.5 minute timeout? How about doing `redeliver` in parallel with `Promise.all`?

That and/or failures from GitHub redelivery API due to too many requests at once. Using `Promise.all()` would make that last one worse. The error handling of large batches of redeliveries is not ideal here.

jackie-linz commented:
> That and/or failures from GitHub redelivery API due to too many requests at once. Using `Promise.all()` would make that last one worse.

I noticed that you are using `@octokit/rest` instead of `octokit`. [octokit](https://github.com/octokit/octokit.js) has built-in throttling and retry implementing all their best practices. It would even handle a `Promise.all` to avoid hitting rate limits.

Was there a reason that it's not used?

> The error handling of large batches of redeliveries is not ideal here.

According to their best practice, mutation requests (`POST`/`PUT`/etc) should have 1s wait in between, so within 4.5 minutes we can send at most 270 redelivery requests.

So yes. the current implementation does not handle total failure scenario where there's more than 200 failures every 5 minutes.

Using a queue is one option - but we still need to take care of the rate limit.

Alternatively, I think it's ok to not retry when there's more than, say, 100 failures found. I think it's reasonable to assume the problem is not transcient in that case.

jackie-linz commented:
@kichik are you able to fix the build so I can download the new build artifact to test?

jackie-linz commented:
Deployed the latest code and performance wise it looks good
<img width="525" height="306" alt="image" src="https://github.com/user-attachments/assets/6688c420-f4f8-45e3-8cea-dbea26e94f02" />

Note I made 2 deployments as indicated by red marker on chart - the second time changing a few of the `debug` log to `info`.

Looks like first execution takes about 10s, with subsequent executions taking around 1-4 seconds.

Will monitor logs to see if redelivery happen as expected

jackie-linz commented:
> That feels weirdly too much? Is that what it took for your original code too? It needs 4 seconds to go over 5 minutes of deliveries?

That's normal, we do have quite a lot of dev work going on. Here's the comparison of the 2 for the same period after deployment

Original
<img width="799" height="449" alt="image" src="https://github.com/user-attachments/assets/3853fc54-8d88-4334-bbc1-428dcd245516" />

New
<img width="797" height="455" alt="image" src="https://github.com/user-attachments/assets/0ed9813f-944e-4e1f-a267-be5d23519017" />

The new have occasional spikes to 10-12 seconds, likely due to cold start

jackie-linz commented:
> > I noticed that you are using `@octokit/rest` instead of `octokit`. [octokit](https://github.com/octokit/octokit.js) has built-in throttling and retry implementing all their best practices. It would even handle a `Promise.all` to avoid hitting rate limits.
> > Was there a reason that it's not used?
> 
> I think there was. But I'm not sure what it was. I feel like it was either bundle size, or maybe something was missing like app authentication.

my original code was implemented with `octokit` locvally and did not run into any problem.

```typescript
import { createAppAuth } from '@octokit/auth-app';
import { Octokit } from 'octokit';

export const handler = async (): Promise<void> => {
  const setting = await getGithubSetting();
  const privateKey = await getGithubPrivateKey();

  const octokit = new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: setting.appId,
      privateKey,
    },
  });
```

> You can uncomment this line to get debug logs:
> 
> https://github.com/jackie-linz/cdk-github-runners/blob/67cc182e77283dde1fb1bc174da869ef643742d9/src/webhook-redelivery.ts#L47

Yes, but that's going to include all the `Processing webhook delivery` logs for each delivery being checked, which would generate a LOT of logs.

Most runs would find 0 failures, and the current logging level means no log would be generated at all. Consider making the 3 loop terminating logs `info` instead of `debug`? i.e.

- Reached last processed delivery ID
- Stopping at old delivery
- No more webhook deliveries to process

And even better, keep a count of the deliveries processed and include it in the log for a bit of stats? here's a patch [logging.patch](https://github.com/user-attachments/files/21337973/logging.patch)

jackie-linz commented:
I can confirm that the new implementation redelivers the failed webhooks.
<img width="1105" height="563" alt="image" src="https://github.com/user-attachments/assets/a758644f-3d78-4294-8bbf-9ce2f5aeb404" />
